### PR TITLE
executor: write multi-valued index

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/40207

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: table/tables/index.go [hunk 1]
```diff
@@ -103,236 +103,299 @@ func (c *index) GenIndexValue(sc *stmtctx.StatementContext, distinct bool, index
...
-		tempKey         []byte
-		keyVer          byte
-		keyIsTempIdxKey bool
-	)
-	if !opt.FromBackFill {
-		key, tempKey, keyVer = GenTempIdxKeyByState(c.idxInfo, key)
-		if keyVer == TempIndexKeyTypeBackfill || keyVer == TempIndexKeyTypeDelete {
-			key, tempKey = tempKey, nil
-			keyIsTempIdxKey = true
+	for _, value := range indexedValues {
```
Comment (xiongjiwei) [inline table/tables/index.go:179]: most of the following code is not changed, just put them into this `for` loop

File: table/tables/index.go [hunk 2]
```diff
@@ -103,236 +103,299 @@ func (c *index) GenIndexValue(sc *stmtctx.StatementContext, distinct bool, index
...
 
 // Delete removes the entry for handle h and indexedValues from KV index.
-func (c *index) Delete(sc *stmtctx.StatementContext, txn kv.Transaction, indexedValues []types.Datum, h kv.Handle) error {
-	key, distinct, err := c.GenIndexKey(sc, indexedValues, h, nil)
-	if err != nil {
-		return err
-	}
+func (c *index) Delete(sc *stmtctx.StatementContext, txn kv.Transaction, indexedValue []types.Datum, h kv.Handle) error {
+	indexedValues := c.getIndexedValue(indexedValue)
+	for _, value := range indexedValues {
```
Comment (xiongjiwei) [inline table/tables/index.go:353]: most of the following code is not changed, just put them into this for loop

**Review by YangKeao — APPROVED** [review]
LGTM

File: util/misc.go [hunk 3]
```diff
@@ -394,10 +394,10 @@ func TLSCipher2String(n uint16) string {
 }
 
 // ColumnsToProto converts a slice of model.ColumnInfo to a slice of tipb.ColumnInfo.
-func ColumnsToProto(columns []*model.ColumnInfo, pkIsHandle bool) []*tipb.ColumnInfo {
+func ColumnsToProto(columns []*model.ColumnInfo, pkIsHandle bool, forIndex bool) []*tipb.ColumnInfo {
```
Comment (qw4990) [inline util/misc.go:397]: Should use `forMVIndex` instead of `ForIndex` here?

File: table/tables/index.go [hunk 4]
```diff
@@ -103,236 +103,299 @@ func (c *index) GenIndexValue(sc *stmtctx.StatementContext, distinct bool, index
 	return tablecodec.GenIndexValuePortal(sc, c.tblInfo, c.idxInfo, c.needRestoredData, distinct, false, indexedValues, h, c.phyTblID, restoredData)
 }
 
+func (c *index) getIndexedValue(indexedValues []types.Datum) [][]types.Datum {
```
Comment (qw4990) [inline table/tables/index.go:106]: If the input values are `[1, '[1, 2]']`, then the output should be `[[1, 1], [1, 2]]`?

File: table/tables/index.go [hunk 5]
```diff
@@ -103,236 +103,299 @@ func (c *index) GenIndexValue(sc *stmtctx.StatementContext, distinct bool, index
...
+			if !c.tblInfo.Columns[c.idxInfo.Columns[i].Offset].FieldType.IsArray() {
+				val = append(val, v)
+			} else {
+				if v.IsNull() {
+					val = append(val, v)
+					jsonIsNull = true
+					jsonIdx++
+					continue
+				}
+				elemCount := v.GetMysqlJSON().GetElemCount()
```
Comment (qw4990) [inline table/tables/index.go:128]: Can a MVIndex have multiple JSON-ARRAY columns like `index idx((cast(a->'$.COL1' as signed array)), (cast(a->'$.COL2' as signed array)))`

Comment (xiongjiwei) [reply table/tables/index.go:128]: no, only 1 JSON-ARRAY is allowed

File: table/tables/index.go (hunk 4 above)
Comment (xiongjiwei) [reply table/tables/index.go:106]: yes!

File: util/misc.go (hunk 3 above)
Comment (xiongjiwei) [reply util/misc.go:397]: I do not get it, just about the name?

File: table/tables/index.go [hunk 6]
```diff
@@ -103,236 +103,299 @@ func (c *index) GenIndexValue(sc *stmtctx.StatementContext, distinct bool, index
...
+				for {
+					// JSON cannot be indexed, if the value is JSON type, it must be multi-valued index.
+					if jsonIdx >= elemCount {
+						goto out
```
Comment (xhebox) [inline table/tables/index.go:132]: Just another option, same effect though:  I think you can also use `break outloop` if you prefer it over `goto`.

**Review by xhebox — APPROVED** [review]
LGTM

File: table/tables/index.go (hunk 6 above)
Comment (xiongjiwei) [reply table/tables/index.go:131]: I prefer `goto` 😁

File: table/tables/index.go (hunk 4 above)
Comment (qw4990) [reply table/tables/index.go:106]: Could you add some comments for the format of the output?

**Review by qw4990 — APPROVED** [review]
LGTM

**Pull request merged**